families/track_and_trade/server/db/records.js

```diff
@@ -0,0 +1,148 @@
+const getOwnerId = getAssociatedAgentId('owners')
+const getCustodianId = getAssociatedAgentId('custodians')
+
+const hasName = name => obj => r.eq(name, getName(obj))
+const hasRecordId = recordId => obj => r.eq(recordId, getRecordId(obj))
+const hasPublicKey = publicKey => obj => r.eq(publicKey, getPublicKey(obj))
+
+// const hasAttribute = getAttr => attr => obj => r.eq(attr, getAttr(obj))
+// const hasName = hasAttribute(getName)
+// const hasRecordId = hasAttribute(getRecordId)
+
+const hasBlock = block => obj => {
+  return r.and(
+    obj('startBlockNum').le(block),
+    obj('endBlockNum').gt(block)
+  )
+}
+
+const queryWithCurrentBlock = query => {
+  return r.table('blocks')
+    .orderBy(r.desc('blockNum'))
+    .nth(0)('blockNum')
+    .do(query)
+}
+
+const getTable = (tableName, block) =>
+      r.table(tableName).filter(hasBlock(block))
+
+const findRecord = recordId => block => {
+  return getTable('records', block)
+    .filter(hasRecordId(recordId))
+    .nth(0)
+}
+
+const findProperty = recordId => propertyName => block => {
+  return getTable('properties', block)
+    .filter(hasRecordId(recordId))
+    .filter(hasName(propertyName))
+    .nth(0)
+}
+
+const getAuthorizedReporterKeys = property => {
+  return getReporters(property)
+    .filter(getAuthorization)
+    .map(getPublicKey)
+    .coerceTo('array')
+}
+
+const getReporter = publicKey => block => {
+  return getTable('agents', block)
+    .filter(hasPublicKey(publicKey))
+    .pluck('name', 'publicKey')
+    .nth(0)
+}
+
+const findReportedValues =
+      recordId => propertyName => dataType => reporterKeys => block => {
+        return getTable('propertyPages', block)
+          .filter(hasRecordId(recordId))
+          .filter(hasName(propertyName))
+          .concatMap(getReportedValues)
+          .map(getUpdate(dataType)(reporterKeys)(block))
+          .orderBy(r.desc('timestamp'))
+          .coerceTo('array')
+      }
+
+const getValue = dataType => value => {
+  return r.branch(
+    r.eq(dataType, 'INT'), value('intValue'),
+    r.eq(dataType, 'STRING'), value('stringValue'),
+    r.eq(dataType, 'FLOAT'), value('floatValue'),
+    r.eq(dataType, 'BYTES'), value('bytesValue'),
+    r.eq(dataType, 'LOCATION'), value('locationValue'),
+    'intValue'
+  )
+}
+
+const getUpdate = dataType => reporterKeys => block => value => {
+  return r.expr({
+    'value': getValue(dataType)(value),
+    'timestamp': value('timestamp'),
+    'reporter': getReporter(reporterKeys.nth(value('reporterIndex')))(block)
+  })
+}
+
+/* Queries */
+
+const fetchPropertyQuery = (recordId, propertyName) => {
+  return queryWithCurrentBlock(block => {
+    return findRecord(recordId)(block).do(record => {
+      return findProperty(recordId)(propertyName)(block).do(property => {
+        return findReportedValues(recordId)(propertyName)(getDataType(property))(getAuthorizedReporterKeys(property))(block).do(values => {
```

Comment:
So, what you can do instead of these nested `do` statements, is just make your queries in an array you return. 
```
return queryWithCurrentBlock(block => {
  return [
    findRecord(recordId)(block),
    findProperty(recordId)(propertyName)(block),
    . . .
  ]
}).do(([ record, property, . . . ]) => {
  return r.expr({
    name: getName(property),
    recordId: getRecordId(record),
    . . .
  })
})
```